cpp/src/quantiles/quantiles.cu

```diff
@@ -1,109 +1,90 @@
-                 rmm::mr::get_default_resource(),
-               cudaStream_t stream = 0)
-    {
-        if (input.size() == input.null_count()) {
-            return std::make_unique<numeric_scalar<ScalarResult>>(0, false, stream);
-        }
+    auto sortmap_lookup = [sortmap]
+        __device__(size_type idx) {
+            return *(sortmap + idx);
+        };
 
-        if (input.size() == 1) {
-            auto result = get_array_value<ScalarResult>(input.begin<T>(), 0);
-            return std::make_unique<numeric_scalar<ScalarResult>>(result, true, stream);
-        }
+    auto quantile_idx_lookup = [sortmap_lookup, interp, size=input.num_rows()]
+        __device__ (double q) {
+            return detail::select_quantile<size_type>(sortmap_lookup,
+                                                      size,
+                                                      q,
+                                                      interp);
+        };
 
-        auto valid_count = input.size() - input.null_count();
+    rmm::device_vector<double> q_device = q;
 
-        if (not column_order.is_ordered)
-        {
-            table_view const in_table { { input } };
-            auto sortmap = sorted_order(in_table, { order::ASCENDING }, { null_order::AFTER });
-            auto sortmap_begin = sortmap->view().begin<size_type>();
-            auto input_begin = input.begin<T>();
+    auto quantile_idx_iter = thrust::make_transform_iterator(q_device.begin(),
+                                                             quantile_idx_lookup);
 
-            auto selector = [&](size_type location) {
-                auto idx = detail::get_array_value<size_type>(sortmap_begin, location);
-                return detail::get_array_value<T>(input_begin, idx);
-            };
+    return detail::gather(input,
+                          quantile_idx_iter,
+                          quantile_idx_iter + q.size(),
+                          false,
+                          false,
+                          false,
+                          mr);
+}
 
-            auto result = select_quantile<ScalarResult>(selector, valid_count, percent, interp);
-            return std::make_unique<numeric_scalar<ScalarResult>>(result, true, stream);
-        }
+}
 
-        auto input_begin = column_order.ordering == order::ASCENDING
-            ? input.begin<T>() + (column_order.null_ordering == null_order::BEFORE ? input.null_count() : 0)
-            : input.begin<T>() - (column_order.null_ordering == null_order::AFTER  ? input.null_count() : 0) + input.size() - 1;
+std::unique_ptr<table>
+quantiles(table_view const& input,
+          std::vector<double> const& q,
+          interpolation interp,
+          cudf::sorted is_input_sorted,
+          std::vector<order> const& column_order,
+          std::vector<null_order> const& null_precedence,
+          rmm::mr::device_memory_resource* mr)
+{
+    CUDF_EXPECTS(interp == interpolation::HIGHER ||
```

Comment:
Could check if any of the columns are non-arithmetic with a `std::any_of` and accept arithmetic interpolation strategies if all are arithmetic.

Reply:
See https://github.com/rapidsai/cudf/issues/4011#issuecomment-582991395